### PR TITLE
Factorization for Integers

### PR DESCRIPTION
**Pull request by Marcel583**
**Description**

This PR implements the Factorization object.

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/utils/factorization/from.rs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        let factor_1 = factors.0.into();
+        let factor_2 = factors.1.into();
+        let mut out = Self::default();
+        unsafe { _fmpz_factor_append(&mut out.factors, &factor_1.value, 1) };
+        unsafe { _fmpz_factor_append(&mut out.factors, &factor_2.value, 1) };
+        out
+    }
+}
+
+impl From<&Factorization> for Vec<(Z, u64)> {
```
Comment (Marvin-Beckmann) [inline src/utils/factorization/from.rs:63]: if this is implemented in one direction, I think it should be implemented in the other as well, i.e. From<&Vec<(Z, u64)>> for Factorization

**Review by Marvin-Beckmann — CHANGES_REQUESTED** [review]
looks godd, just some minor tweaks

File: src/utils/factorization/from.rs [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            out.push((Z::ONE, 1));
+        }
+        out
+    }
+}
+
+#[cfg(test)]
+mod tests_from_one {
+    use crate::utils::Factorization;
+
```
[the remark this answers is not in the document]
Comment (Marcel583) [reply src/utils/factorization/from.rs:106]: done

[the remark this answers is not in the document]
Comment (Marcel583) [reply src/utils/factorization/from.rs:106]: done

File: src/utils/factorization/from.rs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    #[test]
+    fn from_two_unrefined() {
+        let fac = Factorization::from((8, 8));
+
+        assert_eq!("[(8, 1), (8, 1)]", fac.to_string());
+    }
+}
+
+#[cfg(test)]
+mod tests_from_fac {
```
[the remark this answers is not in the document]
Comment (Marcel583) [reply src/utils/factorization/from.rs:171]: done

File: src/utils/factorization/from.rs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+        if out.is_empty() && factors.factors.sign == 1 {
+            out.push((Z::ONE, 1));
+        }
+        out
+    }
+}
+
+#[cfg(test)]
+mod tests_from_one {
```
[the remark this answers is not in the document]
Comment (Marcel583) [reply src/utils/factorization/from.rs:104]: done

File: src/utils/factorization/refine.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    #[test]
+    fn refine_correct_one() {
+        let mut fac = Factorization::from(1);
+
+        fac.refine();
+
+        assert_eq!("[(1, 1)]", fac.to_string());
+    }
+
+    #[test]
```
[the remark this answers is not in the document]
Comment (Marcel583) [reply src/utils/factorization/refine.rs:100]: done

File: src/utils/factorization/from.rs (hunk 1 above)
Comment (Marcel583) [reply src/utils/factorization/from.rs:63]: done

**Review by Marvin-Beckmann — APPROVED** [review]

**Pull request merged**